Make.inc

```diff
@@ -55,42 +56,74 @@ USEIFC  ?= 0
 
 ifeq ($(USE_MKL), 1)
 $(warning "The julia make variable USE_MKL has been renamed to USE_INTEL_MKL")
-USE_INTEL_MKL = 1
+USE_INTEL_MKL := 1
 endif
 
 # libc++ is standard on OS X 10.9, but not for earlier releases
-USE_LIBCPP = 0
+USE_LIBCPP := 0
 
 # assume we don't have LIBSSP support in our compiler, will enable later if likely true
-HAVE_SSP = 0
+HAVE_SSP := 0
 
 # GC debugging options
-WITH_GC_VERIFY = 0
-WITH_GC_DEBUG_ENV = 0
+WITH_GC_VERIFY := 0
+WITH_GC_DEBUG_ENV := 0
 
 # Prevent picking up $ARCH from the environment variables
-ARCH=
+ARCH:=
+
+# pick up BUILDROOT from O= if it isn't already set (from recursive make)
+ifeq ($(BUILDROOT),)
+ifeq ("$(origin O)", "command line")
+  BUILDROOT := $(abspath $O)
+  BUILDDIR := $(abspath $(BUILDROOT)/$(shell $(JULIAHOME)/contrib/relative_path.sh $(JULIAHOME) $(SRCDIR)))
+  $(info $(shell printf '\033[32;1mBuilding into $(BUILDROOT)\033[0m')) # use printf to expand the escape sequences
+else
+  BUILDROOT:=$(JULIAHOME)
+endif
+endif
+export BUILDROOT
+unexport O
+
+# Make sure the user didn't try to specify a path that will confuse the shell / make
+METACHARACTERS := ][?*{}() $$%:;&|!\#,\\`\":
+ifneq (,$(findstring ',$(value BUILDROOT)))
+$(error cowardly refusing to build into directory with a shell-metacharacter in the path)
```

Comment:
fortunately, there's google, which is quite happy to give you a list of these characters (this list is still missing a few like `<` `>` and `\n`, but i think this'll suffice since it doesn't actually seem to have bit anyone yet)